### PR TITLE
Merging the changes from my fork

### PR DESCRIPTION
**Pull request by MatthewL246**
This PR contains all the changes I made in my fork, including Dockerization, functionality to redirect Pretendo requests to a local server, creating HAR network dumps, and an updated readme.

For https://github.com/PretendoNetwork/website/pull/233.

Note that GitHub Actions is used to automatically build the Docker container and push it to the GitHub Container Registry.

### CODE REVIEW

**Comment by jonbarrow** [comment]
Looks good to me 👍 Dani is currently looking into a couple bugs with upstream mitmproxy, and after that we will likely continue work here as well for the docs. I'd like to credit you in the docs, would a GitHub mention be sufficient?

How does this PR look to you @DaniElectra, I know you use mitmproxy as your main proxy?

**Comment by MatthewL246** [comment]
Thanks, a link to my GitHub sounds good.

File: README.md
Comment (MatthewL246) [inline README.md]: I wanted to address this, are the necessary patches in a private repo? Or are users supposed to use the ones included in Inkay/Nimbus?

Comment (jonbarrow) [reply README.md]: Our public patches do disable SSL verification but they also handle redirection, so they aren't suitable for proxy use alone

For the 3DS, a stand-alone SSL patch can be found in the fork used for this PR made by @InternalLoss https://github.com/SciresM/3DS-SSL-Patch/pull/2

For the Wii U, a stand-alone SSL patch is available in a private repo on our org. Truth be told, I don't remember why we went with a private repo? I'll ask around and see if I can figure out why we did that

It might be a good idea to just have the patches in this repo though for easier access, since the only real time you'd want an SSL patch is when using a proxy server. InternalLoss is part of our team now, so I'm sure he wouldn't have an issue with us hosting it here. And the Wii U patch was made by us originally anyway

File: README.md
Comment (jonbarrow) [inline README.md]: This scrubbing is no longer the case. When this was originally being implemented, we made use of the Charles CLI to help with the conversions between multiple archive formats. But it turns out that the Charles CLI will do incomplete conversions, losing information such as multipart form upload boundaries rendering many dumps useless to us

Because of this, we opted to just scrap the feature altogether rather than half-ass it for only some dumps. Instead users dumps are uploaded to a private channel on the server which only 6 people besides myself have access to, all of them being trusted developers (compared to the 23 people currently listed as developers on our Discord). We also have a warning on our new docs page regarding this security issue, so users may opt to not provide dumps if they are not comfortable with it

File: README.md
Comment (jonbarrow) [inline README.md]: > create the dump HAR file

This reads like the HAR has to be created manually? Is that the case? If so, maybe some steps to create the HAR should be added?

**Review by jonbarrow — COMMENTED** [review]
Just a couple notes on stuff I missed earlier

File: README.md
Comment (MatthewL246) [reply README.md]: No, just a wording issue. I meant that the HAR dump is created by mitmproxy when it exits, as in it's not updated live as requests come in.

File: README.md
Comment (MatthewL246) [reply README.md]: Where should users upload the dumps? Should they send them to a public channel or DM them to the bot?

Comment (jonbarrow) [reply README.md]: The bot detects all dumps sent in all channels, deletes the message the user sent, and reuploads them to the private channel alerting the user

File: README.md
Comment (jonbarrow) [reply README.md]: Gotcha :+1:

File: README.md
Comment (jonbarrow) [reply README.md]: While the repo is closed source, I have been known to give out the precompiled patch to people who need it. So for now, I've just uploaded it to Google Drive https://drive.google.com/file/d/14IM--ZcY5cEGNYR30dAgszZKvjjflkuR/view?usp=drive_link

Feel free to download it and just include it in this repo. Same with the IPS patch, I was given permission to just put the patch here in this repo

After this is merged we can deal with the logistics of the private repo and all that on our side

Comment (MatthewL246) [reply README.md]: I can't access the Google Drive link.

Comment (jonbarrow) [reply README.md]: Goes to show how much I use Google Drive. This one should work? https://drive.google.com/file/d/14IM--ZcY5cEGNYR30dAgszZKvjjflkuR/view?usp=sharing

Comment (MatthewL246) [reply README.md]: Got it

Comment (jonbarrow) [reply README.md]: For context this is an Aroma setup module, so it goes in `sd:/wiiu/environments/[ENV]/modules/setup`

File: config-3ds.yaml
Comment (DaniElectra) [inline config-3ds.yaml]: Is it necessary to specify the web host?

File: config-wiiu.yaml
Comment (DaniElectra) [inline config-wiiu.yaml]: Is it necessary to specify the web host?

File: README.md
Comment (DaniElectra) [inline README.md]:
```suggestion
7. Press `Control` and `C` in the terminal window to stop the proxy. This will
```

File: config-local.yaml
Comment (DaniElectra) [inline config-local.yaml]: Not sure if these options are needed for Pretendo redirection? I would leave it to user's choice

**Review by DaniElectra — COMMENTED** [review]
Only some little comments

File: config-3ds.yaml
Comment (MatthewL246) [reply config-3ds.yaml]: I wasn't sure if it made more sense to use mitmweb or mitmdump, but it seems like mitmdump is the simplest way to go for running inside a container.

**Review by DaniElectra — APPROVED** [review]
Looks good to me 👍

**Comment by MatthewL246** [comment]
@jonbarrow It looks like the Docker image build succeeded but trying to pull it causes the error `Error response from daemon: Head "https://ghcr.io/v2/pretendonetwork/mitmproxy-nintendo/manifests/local": unauthorized`. I wonder if there is something in your organization settings making images private?

Update: [This might help](https://docs.github.com/en/packages/learn-github-packages/configuring-a-packages-access-control-and-visibility)

**Comment by jonbarrow** [comment]
Thanks for the update! I also noticed this, I've brought it up in our architecture channel on Discord so we can sort things out (as I'm currently unavailable)

**Comment by jonbarrow** [comment]
@MatthewL246 I've never worked with the GitHub registry before (didn't even know it existed until now). You are correct that this is due to it being private. But it seems as if I am unable to make them public?

I have asked our lead architecture dev @SuperMarioDaBom to take a look at this and see about moving to the official Docker registry, as we have an org there as well, but due to lack of free time I am unsure when this will happen (truth be told, I am personally also fairly unfamiliar with Docker. It's something that's been on my TODO list but never got around to until recently)

EDIT: I believe I got it, @hauntii pointed out there's an org level permission I had to enable

Screenshot from 2024-01-11 15-04-48

**Comment by MatthewL246** [comment]
Moving to the Docker registry sounds totally reasonable. I just chose the GitHub registry because it's convenient to use with GitHub Actions and free for public repos. [I think this post might relate to your issue](https://github.com/orgs/community/discussions/26014).

Edit: Missed your edit, I see that you made it public and I can successfully pull the image now.

**Pull request merged**
